src/System.IO.Ports/src/System/IO/Ports/SerialStream.Unix.cs

```diff
@@ -563,27 +624,231 @@ internal SafeFileHandle OpenPort(string portName)
+                }
+
+                return ret;
+            }
+
+            return 0;
+        }
+
+        private unsafe void IOLoop()
+        {
+            bool eofReceived = false;
+            //bool readyForReceivedEvent = true;
+            long totalBytesRead = 0;
+            // last value of totalBytesRead + BytesToRead
+            long lastTotalBytesAvailable = BytesToRead;
+            while (IsOpen && !eofReceived && !_ioLoopFinished)
+            {
+                // PollEvents will wait if both flags are false
+                Interop.Sys.PollEvents events = PollEvents(1,
+                                                           pollReadEvents: _readQueue.Count > 0,
+                                                           pollWriteEvents: _writeQueue.Count > 0);
+
+                if (events.HasFlag(Interop.Sys.PollEvents.POLLNVAL) ||
+                    events.HasFlag(Interop.Sys.PollEvents.POLLERR))
+                {
+                    // bad descriptor or some other error we can't handle
+                    break;
+                }
+
+                if (events.HasFlag(Interop.Sys.PollEvents.POLLIN))
+                {
+                    int bytesRead = DoIORequest(_readQueue, ProcessRead);
+                    totalBytesRead += bytesRead;
+                }
+
+                // check if there is any new data (either already read or in the driver input)
+                // this event is private and handled inside of SerialPort
+                // which then throttles it with the threshold
+                long totalBytesAvailable = totalBytesRead + BytesToRead;
+                if (totalBytesAvailable > lastTotalBytesAvailable)
+                {
+                    lastTotalBytesAvailable = totalBytesAvailable;
+
+                    // We need new task so that this thread doesn't get deadlocked when someone calls
+                    // Read from within the event
+
+                    Task.Factory.StartNew(s => {
+                            var thisRef = (SerialStream)s;
+                            thisRef.DataReceived(thisRef, new SerialDataReceivedEventArgs(SerialData.Chars));
+                        },
+                        this,
+                        CancellationToken.None,
+                        TaskCreationOptions.DenyChildAttach,
+                        TaskScheduler.Default);
+                }
+
+                if (events.HasFlag(Interop.Sys.PollEvents.POLLOUT))
+                {
+                    DoIORequest(_writeQueue, ProcessWrite);
+                }
+            }
+        }
+
+        private static CancellationToken GetCancellationTokenFromTimeout(int timeoutMs)
+        {
+            return timeoutMs == SerialPort.InfiniteTimeout ?
+                CancellationToken.None :
+                (new CancellationTokenSource(Math.Max(timeoutMs, TimeoutResolution))).Token;
+        }
+
+        class SerialStreamAsyncResult
```

Comment:
This can derive from the TaskCompletionSource instead of containing one.  Then at least we'll have just one allocation instead of two.

Reply:
Good point